File: nxin-messagecenter-client/mc.py
# -*- coding:utf-8 -*-
import sys
import time
import hashlib
import requests
from lxml import etree
from lxml.builder import E


domain = "XXX.nxin.com"
secret = "xxxxx"
sys_id = ""


def gen_xml(message, phones):
    if not isinstance(message, unicode):
        raise TypeError("%s must decode as unicode" % message)
    return etree.tostring(
        E("ShortMessage",
          E("sendSort", "SMS"),
          E("sendType", "COMMON_GROUP"),
          E("isGroup", "1"),
          E("phoneNumber", phones),
          E("isSwitchChannelRetry", "1"),
          E("message", message),
          E("remarks", sys_id)), encoding="utf-8")


def send_sms(message, phones, debug=False):
    api = "http://{0}/message/sendsmsCommonNxin".format(domain)
    timestamp = int(round(time.time() * 1000))
    token = hashlib.md5(secret + str(timestamp)).hexdigest()
    xml = gen_xml(message, phones)
    query = {"message": xml, "timestamp": timestamp, "systemId": sys_id, "accessToken": token}

    if debug:
        s = requests.Session()
        req = requests.Request('POST', api, params=query)
        prepped = req.prepare()
        print("url: %s" % prepped.path_url)
        resp = s.send(prepped)
    else:
        resp = requests.post(api, params=query)
    return resp


def test():
    response = send_sms("python testing 短信".decode('utf-8'), "138******")
    print(response.status_code, response.text)


if __name__ == '__main__':
    message = sys.argv[1].decode(sys.stdin.encoding)
    phone = sys.argv[2]
    resp = send_sms(message, phone)
    sys.stdout.write(resp.text.encode(sys.stdout.encoding))

File: nxin-messagecenter-client/mc.go
package main

import (
	"crypto/md5"
	"encoding/hex"
	"encoding/xml"
	"errors"
	"fmt"
	"io/ioutil"
	"net/http"
	"os"
	"strconv"
	"time"
)

type ShortMessage struct {
	SendSort             string `xml:"sendSort"`
	SendType             string `xml:"sendType"`
	IsSwitchChannelRetry int    `xml:"isSwitchChannelRetry"`
	IsGroup              int    `xml:"isGroup"`
	PhoneNumber          string `xml:"phoneNumber"`
	Message              string `xml:"message"`
	Remarks              string `xml:"remarks"`
}

/* Global variables must define by var */
var api = "http://xxxx.nxin.com/message/sendsmsCommonNxin"
var secret = "xxxx"
var sys_id = ""

func genXml(message string, phones string) string {
	data := &ShortMessage{
		SendSort:             "SMS",
		SendType:             "COMMON_GROUP",
		IsSwitchChannelRetry: 1,
		IsGroup:              1,
		PhoneNumber:          phones,
		Message:              message,
		Remarks:              sys_id,
	}
	bytes, _ := xml.Marshal(data) // []byte
	return string(bytes)
}

func sendSmsMessage(message string, phones string, debug bool) {
	client := http.Client{
		Timeout: time.Duration(5 * time.Second),
	}

	// timestamp
	timeStamp := time.Now().UnixNano() / int64(time.Millisecond)
	timeStampString := strconv.FormatInt(timeStamp, 10)

	// token
	md5Ctx := md5.New()
	md5Ctx.Write([]byte(secret + timeStampString))
	cipherStr := md5Ctx.Sum(nil)
	token := hex.EncodeToString(cipherStr)

	req, err := http.NewRequest("POST", api, nil)
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
	query := req.URL.Query()
	query.Add("message", genXml(message, phones))
	query.Add("timestamp", timeStampString)
	query.Add("systemId", sys_id)
	query.Add("accessToken", token)
	query.Add("businessChannel", "OTHERS")
	req.URL.RawQuery = query.Encode()
	if debug {
		println(req.URL.String())
	}

	resp, err := client.Do(req)
	if err != nil {
		fmt.Println(err)
	}
	defer resp.Body.Close()
	resp_body, _ := ioutil.ReadAll(resp.Body)
	fmt.Println(string(resp_body))
}

func main() {
	args_len := len(os.Args)
	if args_len < 3 || args_len > 4 {
		fmt.Println(errors.New("only support two or three arguments"))
		os.Exit(1)
	}
	message := os.Args[1]
	phones := os.Args[2]
	debug := false
	if args_len == 4 {
		debug = true
	}
	sendSmsMessage(message, phones, debug)
}
